metrics: Represent errors and warnings as counters

vadorovsky opened the pull request:
Add counters representing errors and warnings occuring in
cilium-agent.

Fixes #4268

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5286)
<!-- Reviewable:end -->

vadorovsky committed:
metrics: Represent errors and warnings as a counter

Add a counter representing errors and warnings occuring in
cilium-agent.

Fixes #4268

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -255,6 +257,22 @@ var (
+		Namespace: Namespace,
+		Name:      "errors_total",
+		Help:      "Number of total errors in cilium-agent instances",
+	})
```
I have a couple of proposals here:

What do you think about combining errors/warnings into a single metric, separated by label "level=warning"/"level=error"? Then both warnings and errors could occur in a single graph without additional interaction by the person writing a grafana dashboard.

What do you think about using labels for subsystem as well? Then you could at a glance in one graph see not only how many errors there are (or, in another, warnings), but also see which area of Cilium that they correspond to. If it's, for instance, related to datapath, then maybe you could correlate this with other metrics relating to latency or drops.

vadorovsky replied:
Good idea.

joestringer reviewed (commented):
I have some proposals for the metric below.

joestringer commented on pkg/logging/logging.go:
```diff
@@ -85,6 +88,209 @@ var (
+// metrics. For detauls ebout Entry in Logrus see:
+// https://godoc.org/github.com/sirupsen/logrus#Entry
+type Entry struct {
+	wrappedEntry *logrus.Entry
```
I suspect that if you make this unnamed (ie, just `*logrus.Entry` on a line), it would automatically export functions such as `Panic()`, `Fatal()` and friends, so you wouldn't need to write the boilerplate for those functions that don't need wrapping to do the metrics bits. Did you consider trying this?

joestringer reviewed (approved):
Mostly minor nits/cleanups, otherwise LGTM.

ianvernon commented on pkg/metrics/metrics.go:
```diff
@@ -376,3 +399,25 @@ func DumpMetrics() ([]*models.Metric, error) {
+
+// NOTE(mrostecki): For now errors and warning metrics exist only for Cilium
+// daemon, but support of Prometheus metrics in some other components (i.e.
+// cilium-health) is planned.
```
Can you link to a GH issue for this? E.g., `GH-1234`

vadorovsky replied:
Done

joestringer commented on pkg/metrics/metrics.go:
```diff
@@ -255,6 +257,24 @@ var (
+	// Warnings is the number of warnings in cilium-agent instances
+	Warnings = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: Namespace,
+		Name:      "warnings_total",
```
Did you decide against folding warnings and errors together somewhere? I commented at the start and I see a response from you "good idea", but they seem to still be separate.

If it was considered and rejected, that's fine too - just curious if there was further discussion about this.

vadorovsky replied:
I probably misunderstood you. I thought that you want to keep warrnings and errors separate, but label them by subsystem. Do you want me to introduce an another metric which will include both?

vadorovsky replied:
Discussed offline, done

aanm reviewed (changes requested):
reviewing it, I'll require changes

aanm commented on cilium/cmd/endpoint_labels.go:
```diff
@@ -23,6 +23,7 @@ import (
 	endpointid "github.com/cilium/cilium/pkg/endpoint/id"
 	"github.com/cilium/cilium/pkg/labels"
 	"github.com/cilium/cilium/pkg/labels/model"
+	logginghelpers "github.com/cilium/cilium/pkg/logging/helpers"
```
With this PR https://github.com/cilium/cilium/pull/5447 you no longer need this commit.

vadorovsky replied:
nice, thanks

tgraf replied:
I've just merged #5447 so we can drop this commit

aanm commented on pkg/logging/logging.go:
```diff
@@ -85,6 +88,132 @@ var (
+// NewLogger returns a new instance of Logger with default attributes.
+func NewLogger() *Logger {
+	wrappedLogger := logrus.New()
+	wrappedLogger.Formatter = setupFormatter()
```
logrus accepts hooks. You can remove all the wrappers added and replace it with this:
```
// InitializeDefaultLogger returns a logrus Logger with a custom text formatter.
func InitializeDefaultLogger() *logrus.Logger {
	logger := logrus.New()
	logger.Formatter = setupFormatter()
	logger.Hooks.Add(metrics.MyHook)
	return logger
}
```

in metrics:
```
type MyHooker struct {
}

func (m MyHooker) Levels() []log.Level {
	// return all levels
	return nil
}

func (m MyHooker) Fire(e *log.Entry) error {
	LogLevelsCounter.WithLabelValues(e.Level.String()).Inc()
	return nil
}

var (
	MyHook = MyHooker{}
...
```

aanm reviewed (changes requested):
I also noticed you've encountered a couple of import looks. You can fix the metrics package by moving the `pkg/metrics/status.go` to a different package, perhaps `pkg/metrics/status/status.go`.

tgraf reviewed (commented):
LGTM. I assume we take @aanm's suggestion into account?

vadorovsky committed:
make: Check LogSubsys field in logging entry instances

All logging entry instances have to contain LogSubsys field. This
is required for proper labeling of error/warning Prometheus metric
and helpful for debugging.

Signed-off-by: Michal Rostecki <mrostecki@suse.de>

aanm commented on pkg/metrics/logging_hook.go:
```diff
@@ -0,0 +1,77 @@
+	// daemon, but support of Prometheus metrics in some other components (i.e.
+	// cilium-health - GH-4268) is planned.
+	switch {
+	case components.IsCiliumAgent():
```
This can be removed if you create a constructor for `LoggingHook` and you keep an attribute to know which component belongs to which `LoggingHook`.

aanm commented on pkg/metrics/metrics.go:
```diff
@@ -348,6 +359,11 @@ func init() {
+	MustRegister(ErrorsWarnings)
+
+	// Add logging hook for ErrorWarnings metric
+	logging.DefaultLogger.Hooks.Add(&LoggingHook{})
```
Then, this `logging.DefaultLogger.Hooks.Add(&LoggingHook{})` can be called into the Daemon or Cilium-health main.go
cilium-agent main.go
```
logging.DefaultLogger.Hooks.Add(NewLoggingHook("cilium-agent"))
```
cilium-health

```
logging.DefaultLogger.Hooks.Add(NewLoggingHook("cilium-health"))
```

aanm commented on pkg/metrics/logging_hook.go:
```diff
@@ -0,0 +1,77 @@
+		if err != nil {
+			return fmt.Errorf("log entry cannot be serialized and doesn't contain 'subsys' field")
+		}
+		return fmt.Errorf("log entry doesn't comtain 'subsys' field: %s", serializedEntry)
```
`s/comtain/contain`

aanm commented on pkg/metrics/metrics.go:
```diff
@@ -308,6 +310,15 @@ var (
+	// Errors and warnings
+
+	// ErrorsWarnings is the number of errors and warnings in cilium-agent instances
+	ErrorsWarnings = prometheus.NewCounterVec(prometheus.CounterOpts{
```
The metric name is misleading as it will count for the number of `errors`, `warnings`, `panics` and `fatals` .

vadorovsky replied:
Yes, but `ErrorsWarningsPanicsFatals` seems bit weird for me and I thought that we may consider panics and fatals as specific kind of errors.

raybejjani replied:
Would we ever see the increment for a `panic` or `fatal` ? I think both kill the program inside logrus, and it's unlikely that prometheus would scrape at that instant.

vadorovsky replied:
Good point. I will just remove `panic` and `fatal`.

vadorovsky replied:
done

aanm commented on pkg/metrics/logging_hook.go:
```diff
@@ -0,0 +1,77 @@
+	// cilium-health - GH-4268) is planned.
+	switch {
+	case components.IsCiliumAgent():
+		incErrorWarningMetric(entry)
```
You are ignoring the error returned by this function.

aanm commented on pkg/metrics/logging_hook.go:
```diff
@@ -0,0 +1,77 @@
+	// have 'subsystem' field.
+	iSubsystem, ok := entry.Data[logfields.LogSubsys]
+	if !ok {
+		serializedEntry, err := entry.String()
```
If we are doing a pre-check (following commit) then this can probably be a `panic`. (I don't like panics so I leave for others to give opinion about it)

vadorovsky replied:
@joestringer @tgraf - your thoughts about `panic`? I'm OK with it.

joestringer replied:
I feel like leaving a panic in a regular logging path is asking for trouble.

aanm reviewed (changes requested):
Following up the today's weekly meeting. What's the motivation behind this metric? cc @tgraf or @eloycoto ?

aanm commented on pkg/metrics/logging_hook.go:
```diff
@@ -0,0 +1,85 @@
+	// Pick a metric for the component.
+	switch h.component {
+	case components.CiliumAgentName:
+		metric = ErrorsWarnings
```
This is something that can be part of LoggingHook as well.

This method (`Fire(entry *logrus.Entry)`) will be called for all messages so it needs to be short and fast as possible.

joestringer commented on contrib/scripts/check-logging-subsys-field.sh:
```diff
@@ -0,0 +1,30 @@
+
+set -eu
+
+if git grep -n "log[ ]*= logging\.DefaultLogger*" ':(exclude)test/' ':(exclude)pkg/debugdetection/' | grep -v LogSubsys; then
```
Is it OK to use `git grep` here? Will the container build still work in this case?

vadorovsky replied:
You're right, that might be problematic with building from tarballs.

vadorovsky replied:
done

tgraf commented on pkg/metrics/logging_hook.go:
```diff
@@ -0,0 +1,82 @@
+	if !ok {
+		serializedEntry, err := entry.String()
+		if err != nil {
+			return fmt.Errorf("log entry cannot be serialized and doesn't contain 'subsys' field")
```
Where will these error message show up?

vadorovsky replied:
Directly in stderr https://github.com/sirupsen/logrus/blob/master/entry.go#L135

vadorovsky replied:
It's unfortunate that logrus doesn't print any more details or doesn't use the same entry/logger to print the message. I can figure it out later upstream.

tgraf reviewed (changes requested):
LGTM. Only a minor question.
